File: app/src/main/java/com/jbrunton/mymovies/shared/LoadingViewState.java
package com.jbrunton.mymovies.shared;

import android.support.annotation.DrawableRes;

import com.google.auto.value.AutoValue;

@AutoValue
public abstract class LoadingViewState {
    public enum State {
        OK,
        LOADING,
        ERROR
    }

    public static final LoadingViewState OK_STATE = LoadingViewState.builder()
            .setCurrentState(State.OK)
            .build();

    public static final LoadingViewState LOADING_STATE = LoadingViewState.builder()
            .setCurrentState(State.LOADING)
            .build();

    public boolean showLoadingIndicator() {
        return currentState() == State.LOADING;
    }

    public boolean showError() {
        return currentState() == State.ERROR;
    }

    public boolean showContent() {
        return currentState() == State.OK;
    }

    public abstract State currentState();
    public abstract String errorMessage();
    public abstract @DrawableRes int errorIcon();
    public abstract boolean showTryAgainButton();

    public static Builder builder() {
        return new AutoValue_LoadingViewState.Builder()
                .setErrorIcon(0)
                .setErrorMessage("")
                .setShowTryAgainButton(false);
    }

    public static Builder errorBuilder() {
        return builder()
                .setCurrentState(State.ERROR);
    }

    @AutoValue.Builder
    public abstract static class Builder {
        public abstract Builder setCurrentState(State currentState);
        public abstract Builder setErrorMessage(String errorMessage);
        public abstract Builder setErrorIcon(@DrawableRes int errorIcon);
        public abstract Builder setShowTryAgainButton(boolean showTryAgainButton);

        public abstract LoadingViewState build();
    }
}

File: app/src/main/java/com/jbrunton/mymovies/shared/LoadingViewState.kt
package com.jbrunton.mymovies.shared

import android.view.View
import androidx.annotation.DrawableRes
import com.jbrunton.entities.models.Result

data class LoadingViewState<T>(
        val contentVisibility: Int = View.GONE,
        val loadingIndicatorVisibility: Int = View.GONE,
        val errorCaseVisibility: Int = View.GONE,
        val errorText: String = "",
        @DrawableRes val errorIcon: Int = 0,
        val allowRetryVisibility: Int = View.GONE,
        val contentLoadingState: T
) {
    companion object {
        fun <T> success(viewState: T) = LoadingViewState(
                contentVisibility = View.VISIBLE,
                contentLoadingState = viewState
        )

        fun <T> loading(defaultViewState: T) = LoadingViewState(
                loadingIndicatorVisibility = View.VISIBLE,
                contentLoadingState = defaultViewState
        )

        fun <T> failure(error: LoadingViewStateError, defaultViewState: T) = LoadingViewState(
                errorCaseVisibility = View.VISIBLE,
                errorText = error.message,
                errorIcon = error.errorIcon,
                allowRetryVisibility = if (error.allowRetry) { View.VISIBLE } else { View.GONE },
                contentLoadingState = defaultViewState
        )
    }
}

fun <T> Result<T>.toLoadingViewState(defaultViewState: T): LoadingViewState<T> {
    return when (this) {
        is Result.Success -> {
            LoadingViewState.success(this.value)
        }
        is Result.Loading -> {
            val cachedValue = this.cachedValue
            if (cachedValue == null) {
                LoadingViewState.loading(defaultViewState)
            } else {
                LoadingViewState.success(cachedValue)
            }
        }
        is Result.Failure -> {
            val cachedValue = this.cachedValue
            if (cachedValue == null) {
                val error = this.error
                if (error is LoadingViewStateError) {
                    LoadingViewState.failure(error, defaultViewState)
                } else {
                    throw error
                }
            } else {
                LoadingViewState.success(cachedValue)
            }
        }
    }
}